proposals/0000-modifiers.rst

```diff
@@ -0,0 +1,214 @@
+==============
+
+.. author:: Your name
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/370>`_.
+.. contents::
+
+This proposal introduces a new form of syntax ``%blah`` that defines a *modifier*.
+Modifiers somehow change the meaning of the next token. The ``blah`` is parsed
+and renamed as a type. For now, all modifiers
+will be built in, but we might imagine making an extensible feature later.
+
+As of the writing of this proposal, there will be precisely one modifier: the
+``%1`` or ``%Many`` of linear types. However, `#242`_ proposes *matchability*
+essentially as a postfix modifier; under this proposal, the syntax for
+matchability and multiplicity would be unified.
+
+The main goal of this proposal is to set out a future direction
+for the syntax of modifiers.
+
+Motivation
+----------
+
+The latest form of linear types allows us to write ``Int %1 -> Bool`` to write
+a function that must consume its input ``Int`` precisely once. The fact that
+the function is linear is indicated by the ``%1`` before the ``->``.
+
+A separate proposal `#242`_ describes how the matchability of a function will
+be written postfix, with a ``@``. (The precise meaning of "matchability" is
+irrelevant here.) So, if someone wanted to specify both the multiplicity and
+the matchability of a function, they would write e.g. ``Int %Many -> @Unmatchable Bool``.
+The problme is that we can easily imagine yet more decorations one would want
+to put on an arrow. With multiplicity coming before and matchability afterwards,
+there is now no more room for other modifiers (such as injectivity, or visibility,
+or higher-order role, or others we haven't imagined yet).
+
+In addition, GHC *already* supports some modifier-like constructs, but uses
+a different syntax. Specifically, the term-level operators ``inline`` and
+``oneShot`` are given types like the identity function, but they are really
+modifiers on the variable that comes afterwards. We might imagine making
+their special status more explicit by writing them ``%inline`` or ``%oneShot``.
+
+Proposed Change Specification
+-----------------------------
+
+1. Introduce a new extension ``-XModifiers``.
+
+2. With ``-XModifiers``, introduce modifier syntax on arrows as follows (cf.
+   the Haskell 2010 Report for all BNF syntax, and recall that we use braces
+   to denote "zero or more")::
+
+     type     ::= btype [ {modifier} -> type ]
+     prefix%  ::= '%'    -- only in prefix position
+     modifier ::= prefix% atype
```

Comment:
It still feels strange to say that a modifier is a type. It sounds more like a function `Type -> Type` (or `Expr -> Expr` or `Pat -> Pat` in the other cases), but now that sounds a lot like a Template Haskell splice. Should modifiers maybe be a nicer syntax for running TH splices?

Reply:
That's an interesting way to phrase them, but no. If we imagine that TH supports a larger syntax than Haskell, we could also imagine implementing lots of features via TH that have no syntax in "pure" Haskell. But I would rather keep TH and Haskell aligned with each other -- and to avoid needing the TH machinery to, say, specify a linear type.

And a modifier is not a type. It's parsed/renamed as one, but then interpreted separately.

Reply:
> But I would rather keep TH and Haskell aligned with each other -- and to avoid needing the TH machinery to, say, specify a linear type.

Agreed, but I assumed there would be a modifier-free way of specifying things like linear types (possibly horribly ugly) and that modifiers would act more as a shorthand for the full incantation, which feels more like TH.